### PR TITLE
fix(connect): retry the first probe, and let ssh finish before the killer fires (#858)

### LINKED ISSUE #858: connect fails on the first attempt and works on the second: the wallclock killer pre-empts ssh's own bounds, and an initial probe failure is never retried
## Symptom, as reported

`dot-agent-deck connect <name>` **often** fails on the first invocation and succeeds when run again immediately. Reported by the maintainer against a remote named `dot-agent-deck`:

```
Could not reach remote 'dot-agent-deck': probe exceeded 10s wallclock deadline (laptop-side kill after ssh did not return)
Check your ssh config (`~/.ssh/config`), the host is up, and the network path is open.
```

**Not reproduced on the reporter's network** — the findings below are read from the code, and the mechanism behind "works the second time" is a strong candidate rather than a confirmed cause. See *Still to confirm*.

## Finding 1 — the laptop-side killer fires at HALF the budget its own ssh options permit

`SystemSshExecutor::build_command` (`src/remote.rs:394-403`) configures the probe's ssh with, for `secs` = `probe_timeout_secs()` = **10**:

```
ConnectTimeout=10 · ServerAliveInterval=10 · ServerAliveCountMax=1
```

and its own comment states the consequence:

> Together they bound the wallclock at roughly **2\*secs** in the worst case.

But the outer wall-clock kill is armed at `secs` — 10s — and is what produces the message above (`src/remote.rs:835`). So ssh is deliberately allowed ~20s and our own killer shoots it at 10.

**The killer is documented as a backstop against a remote that accepts the connection and never returns; instead it is the binding constraint.** A connection ssh would have completed at 11–20s is killed as unreachable.

Note what this means for the obvious fix: raising `PROBE_TIMEOUT_SECS_DEFAULT` from 10 to 30 **scales the inconsistency rather than removing it** — `ConnectTimeout=30`, `ServerAliveInterval=30`, killer at 30, ssh allowed ~60. It would fire less often, which will read as a fix, but the killer still pre-empts ssh's own bounds. CLAUDE.md rule 5 records this exact trap for #818: a widened bound that changes nothing structural.

## Finding 2 — an initial probe failure is never retried

`run_connect` already has a retry budget: `MAX_CONNECT_ATTEMPTS = 5` with a fixed backoff. But its own comment scopes it:

> a transport failure on attempt N — a spawn that returned 255, **OR (on a reconnect)** a reachability probe failure — gives up once N reaches `MAX_CONNECT_ATTEMPTS`

So a probe failure is retried only on a **reconnect**, never on the initial connect. The reporter's output corroborates this: it contains **no** `'dot-agent-deck' not reachable yet — retrying… (attempt 2/5)` line, which is what the reconnect path prints (`src/connect.rs:1148`).

This is why the user has to retry by hand: the machinery to retry already exists and does not cover the first attempt.

## Finding 3 — candidate mechanism for "it works the second time"

The connect probe does **not** disable ssh connection multiplexing. `apply_observation_options` (`src/remote.rs:522`) sets `ControlMaster=no` and `ControlPath=none`, but it is applied only when `observation: true` — which `for_observation` sets (`:357`, used by `remote doctor`) and `with_wallclock_timeout` does **not** (`:339`).

So with `ControlMaster auto` + `ControlPersist` in the user's `~/.ssh/config`, the first probe creates the multiplexed master socket; our killer then kills the probe's client, but the master can persist, and the second `connect` reuses it and returns immediately.

**This is not obviously a defect.** Unlike `remote doctor`, which must not create state it is there to read, a persistent master is arguably good for `connect`. It is recorded because it is the most likely answer to "is there a warm-up", and because it means the first invocation may be paying setup costs the second does not.

Ordinary ssh warm-up (DNS/ARP caching, `known_hosts`, a suspended VM waking) sits underneath either way.

## Proposal

Three changes, one PR:

1. **Give the killer headroom over what ssh is allowed.** Arm it at roughly `2 × secs` plus a margin so it is a genuine backstop rather than the binding constraint. With that, the existing 10s `ConnectTimeout` may need no change at all — decide on evidence rather than raising both.
2. **Fold the initial probe failure into the existing retry budget**, so the first attempt gets the same bounded retry a reconnect already gets, with the same `not reachable yet — retrying…` line. This alone would make the reported symptom disappear without touching any bound.
3. **Show progress while the probe runs.** Today it is ten silent seconds followed by a wall of red, which is why this needed a code read to explain. A `Connecting…` indicator naming the phase it is waiting on would make a slow-but-successful connect legible too, not only a failing one.

## Verification — no network needed for 1 and 2

`run_connect` takes `&dyn SshExecutor`, `&dyn ConnectSpawner` and `&dyn ReconnectBackoff`, so both defects are unit-testable against fakes with no ssh and no remote.

`src/connect.rs`'s test module already covers this surface — `reconnect_time_unreachable_is_bounded_not_immediate`, `retry_message_differentiates_drop_from_unreachable`, `clean_exit_does_not_reconnect`, `ctrl_c_exit_is_terminal_no_reconnect`. Per the `reproduce-first` skill, **extend** these with the initial-probe case rather than adding a parallel test that duplicates the setup: this is the same behaviour under a different configuration, not a different mechanism.

For finding 1, assert the relationship rather than the number — the killer's deadline must exceed the worst case its own ssh options permit — so the test keeps holding if the constant changes.

## Still to confirm with the reporter

- `grep -A5 -iE 'controlmaster|controlpersist' ~/.ssh/config` — confirms or kills finding 3.
- `time ssh <target> true`, cold then immediately warm — separates "this link genuinely needs more than 10s" from "our probe/retry logic is wrong".
- `~/.local/state/dot-agent-deck/deck.log` around a failure.

## Note on the progress indicator's one subtlety

`connect` ends by handing the terminal to `ssh -t` (`build_connect_command`, `src/connect.rs:752`). Any spinner must be torn down cleanly before that handoff, and must not draw when stdout is not a TTY — `run_connect_default` already computes `is_tty` for the upgrade nudge (`src/connect.rs:1390`), so follow that existing pattern rather than inventing one.

### PR DESCRIPTION
**Pull request by vfarcic**
Closes #858.

`dot-agent-deck connect <name>` often failed on the first invocation with `probe exceeded 10s wallclock deadline (laptop-side kill after ssh did not return)` and succeeded on an immediate retry. Two independent defects produced that; both are fixed here, plus the issue's third item (progress while the probe runs).

## Reproduced first, as failing tests

Per the `reproduce-first` skill, the first deliverable was a test failing for the user's reason — both defects are reachable through `run_connect`'s injected `&dyn SshExecutor` / `&dyn ConnectSpawner` / `&dyn ReconnectBackoff`, so no network is involved.

Existing tests were **extended**, not duplicated. `UnreachableAfterFirstConnectExecutor` (first probe ok, then always unreachable) became `ScriptedProbeExecutor`, whose per-call script covers both configurations of the same behaviour — the reconnect case `[true, false]` that `reconnect_time_unreachable_is_bounded_not_immediate` already used, and the initial-connect case `[false, …]` this issue is about.

| test | red before the fix |
| --- | --- |
| `connect::tests::initial_probe_unreachable_is_retried_then_connects` (new) | `an initial probe failure must be retried, not fatal: HostUnreachable { name: "prod", detail: "connection refused" }` |
| `connect::tests::initial_probe_unreachable_exhausts_the_budget_with_the_actionable_error` (new) | `every attempt in the budget probed, and no more — left: 1, right: 5` |
| `remote::tests::system_ssh_executor_with_wallclock_timeout_sets_ssh_options` (extended) | `the laptop-side kill (7s) must outlast what its own ssh options permit (14s), or it pre-empts ssh instead of backstopping it` |

Controls: `unanswerable_handshake_stays_fatal_without_naming_versions` gained `backoff.count() == 0`, which pins that widening *when* we retry did not widen *what* we retry; and `reconnect_time_unreachable_is_bounded_not_immediate` still asserts the reconnect give-up is `Ok(255)` after `MAX-1` backoffs.

## The three changes

**1. The wallclock killer now backstops ssh instead of pre-empting it.** `build_command` gives the probe's ssh `ConnectTimeout=secs`, `ServerAliveInterval=secs`, `ServerAliveCountMax=1`, which its own comment says bound ssh's wallclock at roughly `2*secs` — while the laptop-side kill was armed at `secs`. New `remote::wallclock_kill_secs(secs) = 2*secs + 5`, and `SystemSshExecutor::kill_deadline_secs()` derives it, so `run` and `run_capped` arm the kill outside what their own options permit. The test asserts the **relationship** (deadline > `ConnectTimeout + ServerAliveInterval × ServerAliveCountMax`, parsed out of the argv), not the number, so it keeps holding if either constant moves.

**`PROBE_TIMEOUT_SECS_DEFAULT` is unchanged at 10**, deliberately. Raising it to 30 scales both bounds together and leaves the killer still inside ssh's, which is the trap the issue names and CLAUDE.md rule 5 records for #818. With the relationship fixed I have no evidence the default itself is wrong — the reporter's link was never measured — so raising it stays a separate argument with separate evidence. Its doc comment now says so.

**2. An initial probe failure joins the existing retry budget.** `MAX_CONNECT_ATTEMPTS = 5` with backoff already existed but only a *reconnect* fed probe failures into it. New `on_probe_unreachable` covers every attempt and prints the same `'<name>' not reachable yet — retrying…` line. What an exhausted **initial** connect returns is unchanged: the actionable `Could not reach remote…` error, not the reconnect path's "connection lost … giving up" (nothing was lost — no session ever ran). That is the `ProbeRetry::Fatal` arm, gated on a new `session_established` flag; the same flag replaces `attempt == 1` as the upgrade-nudge gate, so a retried initial connect does not silently lose its nudge.

**3. `Connecting…` while the probe blocks.** `Connecting to '<name>'… checking the remote deck`, then `…waiting for the handshake`, drawn around each blocking ssh round-trip and erased the moment it returns — so the teardown before the `ssh -t` handoff holds on every path (success, retry, error) rather than on the ones someone remembered. **Not animated**: no dependency, no redraw thread, nothing to race the handoff — the issue said not to let this grow, so it did not. It draws only when stdio is a terminal; `run_connect_default`'s existing `is_tty` now reads `stdin.is_terminal() && stdout.is_terminal()`, since every surface it gates writes to stdout and waits on a human.

## Each change is load-bearing (reverted one at a time)

| reverted | goes red |
| --- | --- |
| `kill_deadline_secs` → `self.wallclock_timeout` | `system_ssh_executor_with_wallclock_timeout_sets_ssh_options` |
| retry guard back to `attempt > 1 && …` | both `initial_probe_unreachable_*` tests |
| drop the `ProbeRetry::Fatal` arm | `initial_probe_unreachable_exhausts_the_budget_with_the_actionable_error` |
| drop `session_established = true` | `reconnect_time_unreachable_is_bounded_not_immediate` |
| `ConnectProgress::new(false)` | `connect_shows_progress_on_a_tty_and_stays_silent_off_one` |

## Deliberately left alone

- **Finding 3 (ssh `ControlMaster` multiplexing as the "warm-up") is not fixed, and remains an unconfirmed candidate.** The connect probe genuinely does not disable multiplexing — `with_wallclock_timeout` sets `observation: false`, so `apply_observation_options`'s `ControlMaster=no` is not applied — but nobody has confirmed the reporter has it configured, and a persistent master is arguably right for `connect`, unlike `remote doctor` which must not create the state it reads. Pending `grep -A5 -iE 'controlmaster|controlpersist' ~/.ssh/config` from the reporter.
- **`map_probe_ssh_error` / `classify_ssh_error` are untouched**, so `map_probe_ssh_error_keeps_unreachable_host_classification` and `map_probe_ssh_error_classifies_forward_failures` stay green and a forward failure stays fatal on attempt 1.
- **One honest cost of that** (stated because the narrow claim is the true one): the mapper deliberately folds an auth failure and a host-key rejection into `HostUnreachable` — same recourse for the user — so those now enter the retry budget too, costing four extra fast round-trips and ~8s of backoff before the same message. Bounded, identical to what a reconnect has always done, and `remote add` fails on that condition long before `connect` is reached. If it becomes common the fix is a finer classification in the mapper, not a special case in the loop.

## What this does not claim

It does **not** make a first-attempt failure impossible — it makes one retried, so a genuinely-down host now takes longer to say so. The arithmetic, since it is the real cost of change 2: a fast-failing cause (refused, no route, DNS) costs the 4 extra round-trips plus 4x2s of backoff, so ~8 seconds; a host that drops packets is bounded by ssh's own `ConnectTimeout=10`, so ~1 minute; and the ceiling — every attempt hanging ssh past its own bounds all the way to the 25s backstop — is a little over 2 minutes. The retry lines and the progress indicator are what make that legible rather than silent, which is why change 3 belongs in the same PR.

The reporter's network was not reproducible here; the two structural defects were, and are what these tests pin.

## Also verified against the real binary

Not only fakes. A registry entry pointing at `192.0.2.1` (TEST-NET, blackholes) with `DOT_AGENT_DECK_SSH_PROBE_TIMEOUT_SECS=1`, run under an isolated `HOME`:

```
$ dot-agent-deck connect blackhole            # stdout piped
'blackhole' not reachable yet — retrying… (attempt 2/5)
'blackhole' not reachable yet — retrying… (attempt 3/5)
'blackhole' not reachable yet — retrying… (attempt 4/5)
'blackhole' not reachable yet — retrying… (attempt 5/5)
Could not reach remote 'blackhole': ssh: connect to host 192.0.2.1 port 22: Connection timed out
Check your ssh config (`~/.ssh/config`), the host is up, and the network path is open.
```

stdout was **0 bytes** there — the indicator drew nothing off a TTY. The same command under a pty (`script -q -e -c`), with control characters shown:

```
^MConnecting to 'blackhole'… checking the remote deck^M                          ^M'blackhole' not reachable yet — retrying… (attempt 2/5)
```

— drawn while the probe blocks, blanked to column 0, and the retry line written over the cleared row.

## Verification

- `cargo fmt --check` — clean.
- `cargo clippy --workspace --all-targets --features e2e,e2e-live -- -D warnings` — clean (all four flags, per CLAUDE.md rule 2).
- `cargo test-fast` — 4019 passed, 0 failed.
- Scoped tests named above, all in `src/connect.rs` and `src/remote.rs` (fast tier).
- `cargo test-e2e remote_doctor` — the only e2e touching the executor seam I changed (`remote doctor` builds its probe executor from `probe_timeout_secs()` too). `dot-agent-deck connect` itself has no e2e coverage; `tests/CATALOG.md` records that surface as deliberately unit-covered remote-harness territory.
- **Rule 12**: this touches no daemon, protocol, orchestration or hook contract — `connect` is a laptop-side ssh wrapper, and neither the `daemon hello` wire nor its parsing changed. No `PROTOCOL_VERSION` bump, no `.breaking.md`, no cross-version run.

Changelog fragment: `changelog.d/858.bugfix.md`.

### CODE REVIEW

File: src/connect.rs [hunk 1]
```diff
@@ -1333,6 +1525,7 @@ pub fn run_connect<R: BufRead, W: Write>(
                 source,
             }
         })?;
+        session_established = true;
```
[the remark this answers is not in the document]
Comment (vfarcic) [reply src/connect.rs]: Valid, and thank you — this was a real defect I introduced, not a style question. `attempt` was carrying two different budgets at once, and your worst case is exact: probes `[f,f,f,f,t]` came up on attempt 5, and the first `255` then landed in `on_transport_failure`'s exhausted branch, so the session got **zero** reconnects rather than the four PRD #148 and `docs/remote-environments.md` both promise.

Fixed in 8fc2b16 by resetting the counter **once**, at the boundary between the two budgets: the session that has just started becomes attempt 1 of the reconnect budget. The reset is gated on `!session_established`, so it happens exactly once per connect — a flapping session still gets `MAX_CONNECT_ATTEMPTS` spawns in total rather than a fresh budget per reconnect, which is the bound that made the shared counter attractive in the first place. On the ordinary path (probe succeeds first time) `attempt` is already 1, so nothing changes there.

Pinned by `initial_probe_retries_do_not_eat_the_reconnect_budget`, which drives your exact case — probes `[f,f,f,f,t]`, spawns `[255,255,0]` — and asserts exit 0 through 3 spawns and 6 backoffs (4 probe + 2 reconnect). Before the fix it returned `Ok(255)` after 1 spawn and 4 backoffs; reverting just the reset turns it red and leaves the other 31 connect tests green. The docs and changelog said "the same budget", which was the sentence the bug was hiding behind — both now say the budgets are separate and what that guarantees.

File: src/connect.rs [hunk 2]
```diff
@@ -1187,6 +1198,170 @@ fn on_transport_failure(
...
+/// Issue #858: the text of the progress indicator for `phase`. Pure, like
+/// [`retry_message`], so what the user reads can be asserted directly.
+///
+/// It names the phase rather than saying only "Connecting…", because the two
+/// phases fail for different reasons and a user watching one stall for ten
+/// seconds is the person best placed to know which.
+fn progress_line(name: &str, phase: ConnectPhase) -> String {
+    match phase {
+        ConnectPhase::BinaryProbe => format!("Connecting to '{name}'… checking the remote deck"),
+        ConnectPhase::Handshake => format!("Connecting to '{name}'… waiting for the handshake"),
```
[the remark this answers is not in the document]
Comment (vfarcic) [reply src/connect.rs:1222]: Not doing this here, and I want to be precise about why rather than wave it off.

The finding is accurate as far as it goes — `progress_line` does interpolate the name raw. But it is not a new exposure, and hardening only the new line would be theatre. The same string already reaches the same terminal, unescaped, from two shipped printers on the same code path: `retry_message` (`src/connect.rs:1159`, `'{name}' not reachable yet — retrying…`) and `RemoteConnectError::HostUnreachable`'s `Display` (`src/connect.rs:146`). A `connect` to an unreachable host prints all three within a few seconds of each other. Escaping one of the three would leave the surface exactly as exposed while reading, from the diff, as if it had been fixed.

The reason the other two were written that way looks deliberate. Every application of `untrusted_text` in `connect.rs` and `remote.rs` is to **remote-written** bytes — `escape_control_and_bidi(output.stderr)` at `connect.rs:496`, `ssh_error_detail` at `:716`, `scrub_remote_text` at `remote.rs:151`. The boundary being defended is the far end of the ssh connection. The remote *name* is on the near side: it comes from `~/.config/dot-agent-deck/remotes.toml`, written by the operator via `remote add --name`. Whoever can put control characters in that field can also change `host` in the same file — and `connect` then hands the user's terminal to whatever machine that names, over ssh. Repainting a status line is the smaller of those two capabilities by some distance.

So my read is that the registry name is not untrusted input in the sense `untrusted_text` exists for. If the team disagrees, I think that is a legitimate call and the change is small — but it should be one change covering `progress_line`, `retry_message`, `HostUnreachable`'s `Display` (probably by escaping at construction in `map_probe_ssh_error`) and the `pick_remote` listing, argued on the trust boundary rather than tacked onto a timeout/retry bugfix. Happy to do exactly that in a follow-up if you want it; I did not open one because I would be filing an issue for something I have just argued is not a risk. Flag it and I will.

**Comment by vfarcic** [comment]
## Update — Greptile review addressed (8fc2b16)

**P1 (valid, fixed).** The first commit made `attempt` carry two budgets at once: probe retries spent getting the *first* session up counted against the reconnects PRD #148 promises after it drops, so a connect whose probe succeeded on attempt 5 got **zero** reconnects. Fixed by resetting the counter once, at the boundary — the session that just started becomes attempt 1 of the reconnect budget — gated on `!session_established` so a flapping session is still bounded at `MAX_CONNECT_ATTEMPTS` spawns in total. `initial_probe_retries_do_not_eat_the_reconnect_budget` pins it (probes `[f,f,f,f,t]`, spawns `[255,255,0]` → exit 0 through 3 spawns and 6 backoffs; `Ok(255)` after 1 spawn before the fix). The docs and changelog said "the same budget" — the sentence the bug hid behind — and now say the two are separate.

**P2 (not here, reasoning on the thread).** The progress line does print the remote name raw, but so do the two shipped printers beside it on the same path — `retry_message` and `HostUnreachable`'s `Display` — and the name comes from the operator's own `remotes.toml`, whose `host` field is a strictly larger capability. Escaping one of three would read as a fix without being one. Offered as a coherent follow-up across all four call sites if the team wants the registry name treated as untrusted.

## CI note

The two red checks on the previous run were both known and in files this PR does not touch:

- **`build-windows`** (required) failed on `spawn::tests::dispatch_018/019/020` — the fixed-75ms-sleep family of #850, which PR #854 is fixing. Same test flaked once locally under load and passes alone in 4.05s.
- **`e2e-deterministic`** (advisory, not required) failed on `orchestration_remit_001_start…`, the known red on `main`; PR #837 fixes it.

**Review by prageethw — APPROVED** [review]
Verified: the wallclock-kill/ssh-options relationship is asserted structurally (not by magic number), the initial-probe retry correctly separates from the reconnect budget (session_established-gated reset, pinned by `initial_probe_retries_do_not_eat_the_reconnect_budget`), and the fatal-vs-retryable classification stays unwidened (backoff-count-0 control test). Both existing Greptile threads are settled — P1 fixed in 8fc2b16, P2 withdrawn after a sound explanation that the name already reaches the terminal unescaped via two existing printers on the same path. CI is fully green. LGTM.
